### PR TITLE
Support nested dropout autograd

### PR DESCRIPTION
**Pull request by YifanShenSZ**
When the initial version came out, `NestedTensor` was not included in the `CompositeImplicitAutograd` key set, so we had to register dropout_nested to dropout and make it forward-only. Now is the time to improve it!

This pr removes dropout_nested; instead native_dropout_nested is implemented along with native_dropout_backward_nested.

Side change: remove dropout__nested since @cpuhrsch suggested to leave out nested in-place ops for now

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -709,26 +708,19 @@ def test_dropout(self, device, dtype):
...
-                    expect_tensor[j] /= 1.0 - p
-        self.assertEqual(nt, expect)
 
-    # dropout works directly on the underlying buffer memory
-    # so contiguous / noncontiguous does not make any difference
+    @dtypes(torch.float, torch.double)
+    def test_dropout_noncontiguous(self, device, dtype):
+        ntensors = 4
+        nt0 = self.random_nt(device, dtype, ntensors, (4, 4))
+        nt1 = nt0.transpose(-1, -2)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:716]: nit: isn't there a helper function somewhere to make a NT non-contiguous?

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -709,26 +708,19 @@ def test_dropout(self, device, dtype):
...
+        p = 0.3
+        with freeze_rng_state():
+            dropouter = torch.nn.Dropout(p)
+            y0 = dropouter(nt0)
+        with freeze_rng_state():
+            y1 = torch.nn.functional.dropout(nt1, p).transpose(-1, -2)
+        self.assertEqual(y0, y1)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:723]: Useful to test backward for non-contiguous case?

File: test/test_nestedtensor.py [hunk 3]
```diff
@@ -709,26 +708,19 @@ def test_dropout(self, device, dtype):
...
 
-    # dropout works directly on the underlying buffer memory
-    # so contiguous / noncontiguous does not make any difference
+    @dtypes(torch.float, torch.double)
+    def test_dropout_noncontiguous(self, device, dtype):
+        ntensors = 4
+        nt0 = self.random_nt(device, dtype, ntensors, (4, 4))
+        nt1 = nt0.transpose(-1, -2)
+        p = 0.3
+        with freeze_rng_state():
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:718]: Crazy that the tests pass! I expected `freeze_rng_state()` to fail for XLA and possibly CUDA

File: aten/src/ATen/native/nested/NestedTensorBackward.cpp [hunk 4]
```diff
@@ -13,7 +13,12 @@
 namespace at {
 namespace native {
 
-// See Note [nested tensor matmul] TODO in NestedTensorMath.cpp
+Tensor native_dropout_backward_nested(const Tensor& grad, const Tensor& mask, double scale) {
+  Tensor result = grad * mask * scale;
+  return result;
+}
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorBackward.cpp:19]: Do we even need this since it's the same as the CPU implementation?

**Review by jbschlosser — COMMENTED** [review]
Looking much better- thanks! Added a few small comments

File: test/test_nestedtensor.py (hunk 1 above)
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:716]: That helper provides 2 NTs with same entries but different buffers. Here I'm testing 2 NTs with same buffer but different entries.

The idea is that since we redispatch `dropout` to buffer, same-buffer NTs should have same result.

File: test/test_nestedtensor.py (hunk 3 above)
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:718]: What happened to XLA and possibly CUDA?

Comment (jbschlosser) [reply test/test_nestedtensor.py:718]: Idk, I was under the impression from some previous work that `freeze_rng_state()` doesn't work reliably for them

File: aten/src/ATen/native/nested/NestedTensorBackward.cpp (hunk 4 above)
Comment (jbschlosser) [reply aten/src/ATen/native/nested/NestedTensorBackward.cpp:19]: I think 1, maybe rename `native_dropout_backward_cpu` to something a little bit more neutral. I don't see us needing future updates, and I think the ideal state is no new logic needed for NT whenever possible. If we need to split them later on for some reason, best to do it then imo.

File: test/test_nestedtensor.py (hunk 1 above)
Comment (jbschlosser) [reply test/test_nestedtensor.py:716]: Alright no prob - can leave it as-is :)

File: test/test_nestedtensor.py (hunk 3 above)
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:718]: Yes last time I found that for in-place version, CUDA has its own inlined `bernoulli_` instead of the standard one, so CUDA in-place would give different result from others

File: test/test_nestedtensor.py (hunk 2 above)
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:723]: just tried. It would fail because NT mul does not support non-contiguous case

Comment (jbschlosser) [reply test/test_nestedtensor.py:723]: Eek, I think mul should support non-contiguous. We can leave this to a future PR.

**Review by jbschlosser — APPROVED** [review]
LGTM! can leave support of non-contiguous NT for mul to a future PR
